pkg/cover/backend/dwarf.go

```diff
@@ -114,62 +120,74 @@ func makeDWARFUnsafe(target *targets.Target, objDir, srcDir, buildDir string,
 			if module.Name == "" {
 				pcBase = info.textAddr
 			}
-			var data []byte
 			var coverPoints [2][]uint64
 			if target.Arch != targets.AMD64 && target.Arch != targets.ARM64 {
 				coverPoints, err = objdump(target, module)
 			} else if module.Name == "" {
-				data, err = fn.readTextData(module)
+				data, err := fn.readTextData(module)
 				if err != nil {
 					errc <- err
 					return
 				}
 				coverPoints, err = readCoverPoints(target, info, data)
+				if err != nil {
+					errc <- err
+					return
+				}
 			} else {
 				coverPoints, err = fn.readModuleCoverPoints(target, module, info)
 			}
+			if err != nil {
+				errc <- err
+				return
+			}
 			allCoverPoints[0] = append(allCoverPoints[0], coverPoints[0]...)
 			allCoverPoints[1] = append(allCoverPoints[1], coverPoints[1]...)
-			if err == nil && module.Name == "" && len(coverPoints[0]) == 0 {
+			if module.Name == "" && len(coverPoints[0]) == 0 {
 				err = fmt.Errorf("%v doesn't contain coverage callbacks (set CONFIG_KCOV=y on linux)", module.Path)
 			}
 			errc <- err
 		}()
-		ranges, units, err := fn.readTextRanges(module)
-		if err != nil {
-			return nil, err
-		}
 		if err := <-errc; err != nil {
 			return nil, err
 		}
-		allRanges = append(allRanges, ranges...)
-		allUnits = append(allUnits, units...)
 	}
-
-	sort.Slice(allSymbols, func(i, j int) bool {
-		return allSymbols[i].Start < allSymbols[j].Start
-	})
-	sort.Slice(allRanges, func(i, j int) bool {
-		return allRanges[i].start < allRanges[j].start
-	})
 	for k := range allCoverPoints {
 		sort.Slice(allCoverPoints[k], func(i, j int) bool {
 			return allCoverPoints[k][i] < allCoverPoints[k][j]
 		})
 	}
-
-	allSymbols = buildSymbols(allSymbols, allRanges, allCoverPoints)
-	nunit := 0
-	for _, unit := range allUnits {
-		if len(unit.PCs) == 0 {
-			continue // drop the unit
+	var allRanges []*DwarfRange
+	for _, symbol := range allSymbols {
+		for _, r := range symbol.Ranges {
+			allRanges = append(allRanges, &DwarfRange{
+				Symbol: symbol,
+				Start:  r[0],
+				End:    r[1],
+			})
+		}
+	}
+	sort.Slice(allSymbols, func(i, j int) bool {
+		if allSymbols[i].End != allSymbols[j].End {
+			return allSymbols[i].End < allSymbols[j].End
 		}
+		return allSymbols[i].Start > allSymbols[j].Start
```

Comment:
What was the motivation to sort exactly this way?

Reply:
you are right, rg.Symbols are not used to find symbol, it should be removed.